python/ray/llm/tests/batch/gpu/processor/test_sglang_engine_proc.py

```diff
@@ -0,0 +1,136 @@
+import sys
+
+import pytest
+
+import ray
+from ray.llm._internal.batch.processor import ProcessorBuilder
+from ray.llm._internal.batch.processor.sglang_engine_proc import (
+    SGLangEngineProcessorConfig,
+)
+
+
+def test_sglang_engine_processor(gpu_type, model_llama_3_2_216M):
+    config = SGLangEngineProcessorConfig(
+        model_source=model_llama_3_2_216M,
+        engine_kwargs=dict(
+            context_length=8192,
+            tp_size=2,
+            dp_size=2,
+            disable_cuda_graph=True,
+            dtype="half",  # Older GPUs (e.g. T4) don't support bfloat16
+        ),
+        runtime_env=dict(
+            env_vars=dict(
+                RANDOM_ENV_VAR="12345",
+            ),
+        ),
+        accelerator_type=gpu_type,
+        concurrency=4,
+        batch_size=64,
+        max_pending_requests=111,
+        apply_chat_template=True,
+        tokenize=True,
+        detokenize=True,
+    )
+    processor = ProcessorBuilder.build(config)
+    assert processor.list_stage_names() == [
+        "ChatTemplateStage",
+        "TokenizeStage",
+        "SGLangEngineStage",
+        "DetokenizeStage",
+    ]
+
+    stage = processor.get_stage_by_name("SGLangEngineStage")
+    assert stage.fn_constructor_kwargs == {
+        "model": model_llama_3_2_216M,
+        "engine_kwargs": {
+            "context_length": 8192,
+            "tp_size": 2,
+            "dp_size": 2,
+            "disable_cuda_graph": True,
+            "dtype": "half",
+        },
+        "task_type": "generate",
+        "max_pending_requests": 111,
+    }
+
+    runtime_env = stage.map_batches_kwargs.pop("runtime_env")
+    assert "env_vars" in runtime_env
+    assert runtime_env["env_vars"]["RANDOM_ENV_VAR"] == "12345"
+    assert stage.map_batches_kwargs == {
+        "zero_copy_batch": True,
+        "concurrency": 4,
+        "max_concurrency": 4,
+        "accelerator_type": gpu_type,
+        "num_gpus": 4,  # Based on tp_size=2, dp_size=2 in engine_kwargs
+    }
+
+
+def test_generation_model(gpu_type):
+    # OPT models don't have chat template, so we use ChatML template
+    # here to demonstrate the usage of custom chat template.
+    chat_template = """
+{% if messages[0]['role'] == 'system' %}
+    {% set offset = 1 %}
+{% else %}
+    {% set offset = 0 %}
+{% endif %}
+
+{{ bos_token }}
+{% for message in messages %}
+    {% if (message['role'] == 'user') != (loop.index0 % 2 == offset) %}
+        {{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}
+    {% endif %}
+
+    {{ '<|im_start|>' + message['role'] + '\n' + message['content'] | trim + '<|im_end|>\n' }}
+{% endfor %}
+
+{% if add_generation_prompt %}
+    {{ '<|im_start|>assistant\n' }}
+{% endif %}
+    """
+
+    processor_config = SGLangEngineProcessorConfig(
+        model_source="meta-llama/Llama-3.2-1B-Instruct",
```

Comment:
@comaniac @kouroshHakha It seems SGLang doesn't support unsloth models and OPT models. So I hard-code a model path here. We can discuss this later.

Reply:
I'm not sure about OPT, but unsloth Llama is exactly the same as Meta Llama so it seems not make sense to me.

Reply:
I see. It's because downloading the model needs a lot of time. It actually works.